sources/packages/backend/libs/services/src/students-assessments/assessment-sequential-processing.service.ts

```diff
@@ -0,0 +1,182 @@
+   * or past applications will be returned.
+   * @param applicationNumber reference application to be used to determined past and future
+   * applications and the one to be considered as current for the {@link SequencedApplications}.
+   * @param studentId student id.
+   * @param programYearId program id to be considered.
+   * @param entityManager used to execute the commands in the same transaction.
+   * @returns sequenced applications.
+   */
+  private async getSequencedApplications(
+    applicationNumber: string,
+    studentId: number,
+    programYearId: number,
+    entityManager: EntityManager,
+  ): Promise<SequencedApplications> {
+    // Sub query to get the first assessment calculation date for an application to be used for ordering.
+    // All applications versions should be considered, which means 'Overwritten' also.
+    // This will ensure that once the application is calculated for the first time its
+    // order in the sequence will never change.
+    const assessmentDateSubQuery = entityManager
+      .getRepository(StudentAssessment)
+      .createQueryBuilder("assessment")
+      .select("assessment.assessmentDate")
+      .innerJoin("assessment.application", "subQueryApplication")
+      .where(
+        "subQueryApplication.applicationNumber = application.applicationNumber",
+      )
+      .orderBy("assessment.assessmentDate")
+      .limit(1)
+      .getQuery();
+    // Sub query to determined if the assessment has at least one non-declined COE (Required or Completed).
+    // If all the COEs from the assessment are declined some user action will be needed in the application
+    // and this application will not be considered for sequential processing.
+    const existsValidCOE = entityManager
+      .getRepository(DisbursementSchedule)
+      .createQueryBuilder("disbursementSchedule")
+      .select("1")
+      .where("disbursementSchedule.studentAssessment.id = currentAssessment.id")
+      .andWhere("disbursementSchedule.coeStatus != :declinedCOEStatus")
+      .limit(1)
+      .getSql();
+    // Returns past, current, and future applications ordered by the first ever executed assessment calculation.
+    const sequentialApplications = await entityManager
+      .getRepository(Application)
+      .createQueryBuilder("application")
+      .select("application.id", "applicationId")
+      .addSelect("application.applicationNumber", "applicationNumber")
+      .addSelect("application.applicationStatus", "applicationStatus")
+      .addSelect("currentAssessmentOffering.id", "currentAssessmentOfferingId")
+      .addSelect(`(${assessmentDateSubQuery})`, "referenceAssessmentDate")
+      .innerJoin("application.student", "student")
+      .innerJoin("application.programYear", "programYear")
+      .innerJoin("application.currentAssessment", "currentAssessment")
+      .innerJoin("currentAssessment.offering", "currentAssessmentOffering")
```

Comment:
If an application with PIR requested is cancelled, then this inner join will lead to reference application not to part of the retuned list right?

Reply:
True, to ensure that the current one would be returned for this scenario we need to convert the inner to a left join.
For any application different than the reference one the `assessment is not null` will also ensure that the offering id will be there. Does it make sense?

Reply:
Changed and added some comments.